ta/pkcs11/src/pkcs11_token.c

```diff
@@ -877,11 +875,14 @@ enum pkcs11_rc entry_ck_token_initialize(uint32_t ptypes, TEE_Param *params)
 		return PKCS11_CKR_PIN_INCORRECT;
 	}
 
+inited:
+	/* Make sure SO PIN counters are zeroed */
 	token->db_main->flags &= ~(PKCS11_CKFT_SO_PIN_COUNT_LOW |
-				   PKCS11_CKFT_SO_PIN_FINAL_TRY);
+				   PKCS11_CKFT_SO_PIN_FINAL_TRY |
+				   PKCS11_CKFT_SO_PIN_LOCKED |
+				   PKCS11_CKFT_SO_PIN_TO_BE_CHANGED);
```

Comment:
I modified here the logic so that SO stuff is setup in stable state and then USER stuff later in here AND then we call the updated_persistent_db() in line 915

You are correct on SO_PIN_LOCKED but for me it was easier to think that SO flags in this state and USER flags in this state and then we are ready to go.

Thou I am not convinced that it is a good idea to allow SO to be locked out in embedded device ;)

With the above removal or move of updated_persistent_db() I believe it is happening as you said. Eg. now it is called just before returning in all paths where there is a update to be done.

Reply:
resolved